### PR TITLE
Support nonstandard shapes for the Squeeze Op

### PR DESCRIPTION
**Pull request by umangyadav**
Support slice, broadcast and transpose shapes for the squeeze op.

### CODE REVIEW

File: src/include/migraphx/op/squeeze.hpp [hunk 1]
```diff
@@ -37,22 +37,30 @@ struct squeeze
...
         }
         std::vector<std::size_t> new_lens;
+        std::vector<std::size_t> new_strides;
         if(axes.empty())
         {
             std::copy_if(old_lens.begin(),
                          old_lens.end(),
                          std::back_inserter(new_lens),
                          [](auto len) { return len != 1; });
+            for(std::size_t i = 0; i < old_lens.size(); i++)
```
Comment (pfultz2) [inline src/include/migraphx/op/squeeze.hpp:57]: Do `for(auto i:range(old_lens.size()))` instead.

File: src/include/migraphx/op/squeeze.hpp [hunk 2]
```diff
@@ -37,22 +37,30 @@ struct squeeze
...
+            {
+                if(old_lens[i] != 1)
+                {
+                    new_strides.push_back(old_strides[i]);
+                }
```
Comment (pfultz2) [inline src/include/migraphx/op/squeeze.hpp:62]: The `copy_if` from above could be combined with this loop so you would just do:

```cpp
                if(old_lens[i] != 1)
                {
                    new_strides.push_back(old_strides[i]);
                    new_lens.push_back(old_lens[i]);
                }
```

File: src/include/migraphx/op/squeeze.hpp (hunk 1 above)
Comment (umangyadav) [reply src/include/migraphx/op/squeeze.hpp]: Done. Thanks.

File: src/include/migraphx/op/squeeze.hpp (hunk 2 above)
Comment (umangyadav) [reply src/include/migraphx/op/squeeze.hpp:59]: Done thanks.

**Review by scxiao — APPROVED** [review]

**Review by turneram — APPROVED** [review]

File: test/ref_ops_nonstd_shape_test.cpp [hunk 3]
```diff
@@ -55,4 +55,62 @@ TEST_CASE(argmin_test_nonstd_shape)
...
+    auto l0 = mm->add_literal(migraphx::literal{s1, data});
+    auto l0_trans =
+        mm->add_instruction(migraphx::make_op("transpose", {{"permutation", {1, 2, 3, 0, 4}}}), l0);
+    mm->add_instruction(migraphx::make_op("squeeze"), l0_trans);
+    auto p_uncompiled = p;
+    p.compile(migraphx::ref::target{});
+    auto result         = p.eval({}).back();
+    auto* mm_uncompiled = p_uncompiled.get_main_module();
+    // ref implementation always does the auto_contiguous pass
+    migraphx::run_passes(*mm_uncompiled, {migraphx::auto_contiguous{}});
```
Comment (scxiao) [inline test/ref_ops_nonstd_shape_test.cpp:73]: I think adding a contiguous here does not test what you want. In this case, the output of the uncompiled program is a transposed shape, you need to manually do a transpose for the output according to the remaining dimension in the above transpose op (permutation is {1, 2, 0}).  and compared the output to that of the compiled program.

**Review by scxiao — COMMENTED** [review]
I think the transposed input maybe incorrect after removing the standard input shape requirement.

**Comment by scxiao** [comment]
@causten Please hold merging this PR since we can improve the unit test. I will let  you know after it is fixed.

**Review by pfultz2 — APPROVED** [review]

File: test/ref_ops_nonstd_shape_test.cpp (hunk 3 above)
Comment (pfultz2) [reply test/ref_ops_nonstd_shape_test.cpp:73]: `contiguous` will preserve order of elements but does not preserve memory layout. However, this test only checks the shape and not the values in the tensor, which is a problem.

File: test/ref_ops_nonstd_shape_test.cpp [hunk 4]
```diff
@@ -55,4 +55,62 @@ TEST_CASE(argmin_test_nonstd_shape)
...
+    auto expected_result = p_uncompiled.eval({}).back();
+    EXPECT(result.get_shape() == expected_result.get_shape());
```
Comment (pfultz2) [inline test/ref_ops_nonstd_shape_test.cpp:75]: You need to check the values are the same: `EXPECT(result == expected_result)`

**Review by pfultz2 — CHANGES_REQUESTED** [review]

File: test/ref_ops_nonstd_shape_test.cpp (hunk 4 above)
Comment (umangyadav) [reply test/ref_ops_nonstd_shape_test.cpp]: Made the change.

File: test/ref_ops_nonstd_shape_test.cpp (hunk 3 above)
Comment (pfultz2) [reply test/ref_ops_nonstd_shape_test.cpp:77]: If you add `auto_contiguous` then you need to add `eliminate_contiguous` to make sure it wont remove a `contiguous` thats needed. Although, we shouldn't be running `auto_contiguous` on the uncompiled version.

Comment (umangyadav) [reply test/ref_ops_nonstd_shape_test.cpp]: @scxiao @pfultz2 I've added a contiguous op instead of using `auto_contiguous` pass.

**Review by scxiao — APPROVED** [review]

File: test/ref_ops_nonstd_shape_test.cpp [hunk 5]
```diff
@@ -55,4 +55,71 @@ TEST_CASE(argmin_test_nonstd_shape)
...
 
+TEST_CASE(squeeze_transpose_test)
+{
+    migraphx::program p;
+    auto* mm                = p.get_main_module();
+    std::vector<float> data = {
+        1.2255,  1.6834,  -2.0305, -0.3221, 0.4701,  0.2583, 0.7545,  2.5758,  -1.6849,
+        0.0928,  0.9022,  -0.8765, -0.4090, 0.9301,  2.0724, -1.5706, 0.4867,  -0.1493,
+        0.6957,  -0.2179, 0.7142,  0.7177,  0.0183,  1.3497, 1.2255,  1.6834,  -2.0305,
+        -0.3221, 1.2255,  1.6834,  -2.0305, -0.3221, 1.2255, 1.6834,  -2.0305, -0.3221};
```
Comment (pfultz2) [inline test/ref_ops_nonstd_shape_test.cpp:66]: Just use `generate_literal` instead.

File: test/ref_ops_nonstd_shape_test.cpp [hunk 6]
```diff
@@ -55,4 +55,71 @@ TEST_CASE(argmin_test_nonstd_shape)
...
+        mm->add_instruction(migraphx::make_op("transpose", {{"permutation", {1, 2, 3, 0, 4}}}), l0);
+    mm->add_instruction(migraphx::make_op("squeeze"), l0_trans);
+    auto p_uncompiled = p;
+    p.compile(migraphx::ref::target{});
+    auto result          = p.eval({}).back();
+    auto expected_result = p_uncompiled.eval({}).back();
+    // contiguous is required to read the values in standard shaped order
+    auto tr_op               = migraphx::make_op("contiguous");
+    auto std_expected_result = tr_op.compute(result.get_shape(), {expected_result});
+    EXPECT(result == std_expected_result);
```
Comment (pfultz2) [inline test/ref_ops_nonstd_shape_test.cpp:79]: Can you add check for the shape so we can see what the final shape should be in the test? Something like:

```cpp
EXPECT(result.get_shape() == migraphx::shape{migraphx::shape::float_type, {4, 3, 3}})`
```

File: test/ref_ops_nonstd_shape_test.cpp (hunk 5 above)
Comment (umangyadav) [reply test/ref_ops_nonstd_shape_test.cpp]: Done.

File: test/ref_ops_nonstd_shape_test.cpp (hunk 6 above)
Comment (umangyadav) [reply test/ref_ops_nonstd_shape_test.cpp:67]: Done.

**Review by pfultz2 — APPROVED** [review]

**Pull request merged**